### PR TITLE
feat: change foreign layer error message at repository.Fetch()

### LINKED ISSUE #319: feat: improve the error message if non-distributable(foreign) layer is not supported
Some artifacts(e.g., Windows base images) contain layers whose distribution is restricted by license. When these images are pushed to a registry, restricted artifacts are not included.

For now, `oras.Copy` and `oras.Fetch` will fail if the target artifact contains an non-distributable(foreign) layer since the layer blob doesn't exist in the remote registry. We should
1) By default skip blob copying if a layer is non-distributable
2) Provide option to enforce copying a non-distributable layer and follows the foreign link for fetching

### PR DESCRIPTION
**Pull request by wangxiaoxuan273**
Resolves #319

Current result with oras-cli:
```
$ oras copy docker.io/library/hello-world@sha256:314cc0309465c7be2936df274b258c843f5a49eb21f63b719d60b51564070b8f wxxacr.azurecr.io/temp439:test00

$ Error: the artifact with foreign layer sha256:5ead999142ecce15e02523c49706a633fa708374d94bb65a254e3a3c117d609b is not supported: sha256:5ead999142ecce15e02523c49706a633fa708374d94bb65a254e3a3c117d609b: not found
```

Signed-off-by: wangxiaoxuan273 <wangxiaoxuan119@gmail.com>

### CODE REVIEW

File: errdef/errors.go [hunk 1]
```diff
@@ -27,4 +27,5 @@ var (
 	ErrSizeExceedsLimit   = errors.New("size exceeds limit")
 	ErrUnsupported        = errors.New("unsupported")
 	ErrUnsupportedVersion = errors.New("unsupported version")
+	ErrForeignLayer       = errors.New("foreign layer is not supported")
```
Comment (wangxiaoxuan273) [inline errdef/errors.go:30]: Need advice for a better message that matches the error level.

File: registry/remote/repository.go [hunk 2]
```diff
@@ -614,6 +615,9 @@ type blobStore struct {
 
 // Fetch fetches the content identified by the descriptor.
 func (s *blobStore) Fetch(ctx context.Context, target ocispec.Descriptor) (rc io.ReadCloser, err error) {
+	if isForeignLayer(target) {
+		return nil, fmt.Errorf("%s: %w", target.Digest, errdef.ErrForeignLayer)
+	}
```
Comment (wangxiaoxuan273) [inline registry/remote/repository.go:620]: Should this check be here, or inside its upstream caller (copy)? It affects what error message to display.

File: registry/remote/repository.go [hunk 3]
```diff
@@ -614,6 +615,9 @@ type blobStore struct {
 
 // Fetch fetches the content identified by the descriptor.
 func (s *blobStore) Fetch(ctx context.Context, target ocispec.Descriptor) (rc io.ReadCloser, err error) {
+	if isForeignLayer(target) {
+		return nil, fmt.Errorf("%s: %w", target.Digest, errdef.ErrForeignLayer)
```
[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply registry/remote/repository.go:619]: Impossible :)

[the remark this answers is not in the document]
Comment (wangxiaoxuan273) [reply registry/remote/repository.go:619]: After offline discussion, we improved the error message and resolved this comment.

File: registry/remote/repository.go (hunk 2 above)
Comment (wangxiaoxuan273) [reply registry/remote/repository.go:620]: Put it here is fine.

File: errdef/errors.go (hunk 1 above)
Comment (wangxiaoxuan273) [reply errdef/errors.go:30]: Changed accordingly, after offline discussion.

File: registry/remote/repository.go [hunk 4]
```diff
@@ -1428,3 +1432,14 @@ func generateIndex(manifests []ocispec.Descriptor) (ocispec.Descriptor, []byte,
...
+
+func isForeignLayer(desc ocispec.Descriptor) bool {
+	switch desc.MediaType {
+	case ocispec.MediaTypeImageLayerNonDistributable,
+		ocispec.MediaTypeImageLayerNonDistributableGzip,
+		ocispec.MediaTypeImageLayerNonDistributableZstd,
+		docker.MediaTypeForeignLayer:
+		return true
+	}
+	return false
```
Comment (Wwwsylvia) [inline registry/remote/repository.go:1444]: nit:
```suggestion
	default:
	    return false
	}
```

File: registry/remote/repository.go (hunk 2 above)
Comment (Wwwsylvia) [reply registry/remote/repository.go:620]: It's the right place. We should also apply the same logic to `blobStore.FetchReference()`.

File: errdef/errors.go (hunk 1 above)
Comment (Wwwsylvia) [reply errdef/errors.go]: I'm wondering would this specific `ErrForeignLayer` be checked by consumers (such as oras and notation)? If not, maybe we can just reuse `ErrUnsupported`.

File: registry/remote/repository.go (hunk 2 above)
Comment (Wwwsylvia) [reply registry/remote/repository.go:620]: Besides, can we have some small unit tests?

**Review by Wwwsylvia — CHANGES_REQUESTED** [review]

File: registry/remote/repository.go (hunk 2 above)
Comment (wangxiaoxuan273) [reply registry/remote/repository.go:620]: Moved it down after getting 404.

File: registry/remote/repository.go [hunk 5]
```diff
@@ -614,6 +615,9 @@ type blobStore struct {
 
 // Fetch fetches the content identified by the descriptor.
 func (s *blobStore) Fetch(ctx context.Context, target ocispec.Descriptor) (rc io.ReadCloser, err error) {
+	if isForeignLayer(target) {
+		return nil, fmt.Errorf("the artifact with foreign layer %s is %w", target.Digest, errdef.ErrUnsupported)
```
Comment (Wwwsylvia) [inline registry/remote/repository.go:619]: Actually, I think "artifact with foreign layer" looks a bit strange for the `blobStore.Fetch` method... 🤔
What if
```suggestion
		return nil, fmt.Errorf("%s: %s: foreign layer is unsupported: %w", target.Digest, target.MediaType, errdef.ErrUnsupported)
```
？

**Comment by qweeah** [comment]
Some repositories may contain foreign layers which pointing to the repository itself, e.g.
```
> oras manifest fetch mcr.microsoft.com/windows/servercore:1607-KB5019964 --pretty --platform windows/amd64
{
  "schemaVersion": 2,
  "mediaType": "application/vnd.docker.distribution.manifest.v2+json",
  "config": {
    "mediaType": "application/vnd.docker.container.image.v1+json",
    "size": 787,
    "digest": "sha256:f443e6ff891920258e9c0ce66f7f5472793b26691ea18d7fc87af0472c677430"
  },
  "layers": [
    {
      "mediaType": "application/vnd.docker.image.rootfs.foreign.diff.tar.gzip",
      "size": 4069985900,
      "digest": "sha256:3889bb8d808bbae6fa5a33e07093e65c31371bcf9e4c38c21be6b9af52ad1548",
      "urls": [
        "https://mcr.microsoft.com/v2/windows/servercore/blobs/sha256:3889bb8d808bbae6fa5a33e07093e65c31371bcf9e4c38c21be6b9af52ad1548"
      ]
    },
    {
      "mediaType": "application/vnd.docker.image.rootfs.foreign.diff.tar.gzip",
      "size": 2213621320,
      "digest": "sha256:29226bae1723379e75abf7d7da45bbf67d5d730611c58bf86df104245db30204",
      "urls": [
        "https://mcr.microsoft.com/v2/windows/servercore/blobs/sha256:29226bae1723379e75abf7d7da45bbf67d5d730611c58bf86df104245db30204"
      ]
    }
  ]
}
```
So it's better to return unsupported **only** when a layer is foreign and not found

File: registry/remote/repository.go (hunk 5 above)
Comment (qweeah) [reply registry/remote/repository.go:619]: The default media type of a blob cannot be foreign, there MUST be an artifact whose manifest is pointing to the to-be-fetched layer.

File: registry/remote/repository.go (hunk 4 above)
Comment (wangxiaoxuan273) [reply registry/remote/repository.go:1444]: changed.

File: registry/remote/repository.go (hunk 2 above)
Comment (wangxiaoxuan273) [reply registry/remote/repository.go]: After discussion, removed the check inside `Fetch()`.

File: internal/descriptor/descriptor.go [hunk 6]
```diff
@@ -45,3 +46,15 @@ func FromOCI(desc ocispec.Descriptor) Descriptor {
 		Size:      desc.Size,
 	}
 }
+
+func IsForeignLayer(desc ocispec.Descriptor) bool {
```
Comment (shizhMSFT) [inline internal/descriptor/descriptor.go:50]: nit: Missing documentation for this method.

File: copy.go [hunk 7]
```diff
@@ -258,6 +259,9 @@ func copyGraph(ctx context.Context, src content.ReadOnlyStorage, dst content.Sto
 func doCopyNode(ctx context.Context, src content.ReadOnlyStorage, dst content.Storage, desc ocispec.Descriptor) error {
 	rc, err := src.Fetch(ctx, desc)
 	if err != nil {
+		if errors.Is(err, errdef.ErrNotFound) && descriptor.IsForeignLayer(desc) {
```
Comment (shizhMSFT) [inline copy.go:262]: Can we do other way around?
```suggestion
			if descriptor.IsForeignLayer(desc) && errors.Is(err, errdef.ErrNotFound) {
```

File: content.go [hunk 8]
```diff
@@ -273,6 +274,9 @@ func Fetch(ctx context.Context, target ReadOnlyTarget, reference string, opts Fe
 		}
 		rc, err := target.Fetch(ctx, desc)
 		if err != nil {
+			if errors.Is(err, errdef.ErrNotFound) && descriptor.IsForeignLayer(desc) {
```
Comment (shizhMSFT) [inline content.go:277]: By analyzing, we don't need to consider foreign layers in `oras.Fetch` as we cannot fetch a foreign layer from a reference string but a descriptor.

Let's revert the changes in the `content.go`.

File: internal/descriptor/descriptor.go (hunk 6 above)
Comment (wangxiaoxuan273) [reply internal/descriptor/descriptor.go:50]: Added the documentation.

File: content.go (hunk 8 above)
Comment (wangxiaoxuan273) [reply content.go:277]: Reverted.

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Review by Wwwsylvia — APPROVED** [review]
LGTM
